src/MetaMorpho.sol

```diff
@@ -341,55 +341,56 @@ contract MetaMorpho is ERC4626, ERC20Permit, Ownable2Step, Multicall, IMetaMorph
-        for (uint256 i; i < withdrawn.length; ++i) {
-            MarketAllocation memory allocation = withdrawn[i];
+        for (uint256 i; i < allocations.length; ++i) {
+            MarketAllocation memory allocation = allocations[i];
             Id id = allocation.marketParams.id();
 
-            if (allocation.marketParams.loanToken != asset()) revert ErrorsLib.InconsistentAsset(id);
+            MORPHO.accrueInterest(allocation.marketParams);
 
-            // Guarantees that unknown frontrunning donations can be withdrawn, in order to disable a market.
-            uint256 shares;
-            if (allocation.assets == type(uint256).max) {
-                shares = MORPHO.supplyShares(id, address(this));
+            Market memory market = MORPHO.market(id);
 
-                allocation.assets = 0;
-            }
+            uint256 supplyShares = MORPHO.supplyShares(id, address(this));
+            uint256 supplyAssets = supplyShares.toAssetsDown(market.totalSupplyAssets, market.totalSupplyShares);
+            uint256 withdrawn = supplyAssets.zeroFloorSub(allocation.assets);
 
-            (uint256 withdrawnAssets, uint256 withdrawnShares) =
-                MORPHO.withdraw(allocation.marketParams, allocation.assets, shares, address(this), address(this));
+            if (withdrawn > 0) {
+                if (allocation.marketParams.loanToken != asset()) revert ErrorsLib.InconsistentAsset(id);
 
-            totalWithdrawn += withdrawnAssets;
+                // Guarantees that unknown frontrunning donations can be withdrawn, in order to disable a market.
+                uint256 shares;
+                if (allocation.assets == 0) {
+                    shares = supplyShares;
+                    withdrawn = 0;
+                }
 
-            emit EventsLib.ReallocateWithdraw(_msgSender(), id, withdrawnAssets, withdrawnShares);
-        }
+                (uint256 withdrawnAssets, uint256 withdrawnShares) =
+                    MORPHO.withdraw(allocation.marketParams, withdrawn, shares, address(this), address(this));
 
-        uint256 totalSupplied;
-        for (uint256 i; i < supplied.length; ++i) {
-            MarketAllocation memory allocation = supplied[i];
-            Id id = allocation.marketParams.id();
-            uint256 supplyCap = config[id].cap;
+                emit EventsLib.ReallocateWithdraw(_msgSender(), id, withdrawnAssets, withdrawnShares);
 
-            if (supplyCap == 0) revert ErrorsLib.UnauthorizedMarket(id);
+                totalWithdrawn += withdrawnAssets;
+            } else {
+                uint256 supplied = allocation.assets.zeroFloorSub(supplyAssets);
+                if (supplied == 0) continue;
 
-            (uint256 suppliedAssets, uint256 suppliedShares) =
-                MORPHO.supply(allocation.marketParams, allocation.assets, 0, address(this), hex"");
+                uint256 supplyCap = config[id].cap;
+                if (supplyCap == 0) revert ErrorsLib.UnauthorizedMarket(id);
```

Comment:
I know that these are details, but I think that it's worth growing these habits. 

The `supplyCap == 0` means two things:
1) explicit: the curator has disabled the market, and you cannot supply to it (or the market has never been enabled)
2) implicit the market that you are trying to supply to have not the same asset as the Metamorpho vault

The first one is easy to see, the second one not. A dev comment would be nice

Reply:
Are you suggesting we should we should revert differently depending on the case?

Reply:
I added a comment as suggested

Reply:
> Are you suggesting that we should revert differently depending on the case?

this depends mostly on you and what you want to monitor. the function is authed so I assume that the person should be trusted. 

it would be needed if you want to monitor an authed user that is trying to supply to a not authorized (asset different from the vault) market. but you have to be very stupid to do that knowing that it would anyway revert